test/common/event/scaled_range_timer_manager_test.cc

```diff
@@ -0,0 +1,450 @@
+
+  // Advance time to T = 2, which should make timers[0] hit its scaled max.
+  simTime().advanceTimeAsync(std::chrono::seconds(1));
+  dispatcher_.run(Dispatcher::RunType::Block);
+
+  // At 4x speed, timers[1] will fire in only 1 second.
+  manager.setScaleFactor(0.25);
+
+  // Advance time to T = 3, which should make timers[1] hit its scaled max.
+  simTime().advanceTimeAsync(std::chrono::seconds(1));
+  dispatcher_.run(Dispatcher::RunType::Block);
+
+  // Advance time to T = 6, which is the minimum required for timers[2] to fire.
+  simTime().advanceTimeAsync(std::chrono::seconds(3));
+  dispatcher_.run(Dispatcher::RunType::Block);
+
+  manager.setScaleFactor(0);
+  // With a scale factor of 0, timers[2] should be ready to be fired immediately.
+  dispatcher_.run(Dispatcher::RunType::Block);
+
+  EXPECT_THAT(*timers[0].trigger_times, ElementsAre(T + std::chrono::seconds(2)));
+  EXPECT_THAT(*timers[1].trigger_times, ElementsAre(T + std::chrono::seconds(3)));
+  EXPECT_THAT(*timers[2].trigger_times, ElementsAre(T + std::chrono::seconds(6)));
+}
+
+TEST_F(ScaledRangeTimerManagerTest, MultipleTimersSameTimes) {
+  ScaledRangeTimerManager manager(dispatcher_, 1.0);
+  std::vector<TrackedTimer> timers;
+
+  const MonotonicTime T = simTime().monotonicTime();
+
+  for (int i = 0; i < 3; ++i) {
+    timers.emplace_back(manager, simTime());
+    timers[i].timer->enableTimer(std::chrono::seconds(1), std::chrono::seconds(2));
+  }
+
+  simTime().advanceTimeAsync(std::chrono::seconds(1));
+  dispatcher_.run(Dispatcher::RunType::Block);
+
+  simTime().advanceTimeAsync(std::chrono::seconds(1));
+  dispatcher_.run(Dispatcher::RunType::Block);
+
+  EXPECT_THAT(*timers[0].trigger_times, ElementsAre(T + std::chrono::seconds(2)));
+  EXPECT_THAT(*timers[1].trigger_times, ElementsAre(T + std::chrono::seconds(2)));
+  EXPECT_THAT(*timers[2].trigger_times, ElementsAre(T + std::chrono::seconds(2)));
+}
+
+TEST_F(ScaledRangeTimerManagerTest, MultipleTimersSameTimesFastClock) {
+  ScaledRangeTimerManager manager(dispatcher_, 1.0);
+  std::vector<TrackedTimer> timers;
+
+  const MonotonicTime T = simTime().monotonicTime();
+
+  for (int i = 0; i < 3; ++i) {
+    timers.emplace_back(manager, simTime());
+    timers[i].timer->enableTimer(std::chrono::seconds(1), std::chrono::seconds(2));
+  }
+
+  simTime().advanceTimeAsync(std::chrono::seconds(1));
+  dispatcher_.run(Dispatcher::RunType::Block);
+  // The clock runs fast here before the dispatcher gets to the timer callbacks.
+  simTime().advanceTimeAsync(std::chrono::seconds(2));
+  dispatcher_.run(Dispatcher::RunType::Block);
+
+  EXPECT_THAT(*timers[0].trigger_times, ElementsAre(T + std::chrono::seconds(3)));
+  EXPECT_THAT(*timers[1].trigger_times, ElementsAre(T + std::chrono::seconds(3)));
+  EXPECT_THAT(*timers[2].trigger_times, ElementsAre(T + std::chrono::seconds(3)));
+}
+
+TEST_F(ScaledRangeTimerManagerTest, ScheduledWithScalingFactorZero) {
+  ScaledRangeTimerManager manager(dispatcher_, 0.0);
+
```

Comment:
Please add a test that covers use of setScaleFactor.  There seems to be no tests that exercise that function.
Also, prefer use of setScaleFactor instead of passing in a scale value to the constructor.

I recommend exercising some some in-between scale factors like 0.5 with timers that are expected to fire with different orders when scale = 1.0 and scale = 0.5
Cover both the case of timers added before the scale change and timers added after the scale change.  Bonus points for mixing the two.

Reply:
MultipleTimersWithScaling does some of this. I've added some more tests.